docs/source/concepts/ledger-model/ledger-integrity.rst

```diff
@@ -55,26 +61,148 @@ below depicts the resulting order on the paint offer example:
+**contract state** at any point in a given transaction.
+The contract state is changed by creating the contract and by exercising it consumingly.
+At any point in a transaction, we can then define the latest state change in the obvious way.
+Then, given a point in a transaction, the contract state of `c` is:
+
+#. **active**, if the latest state change of `c` was a create;
+
+#. **archived**, if the latest state change of `c` was a consuming exercise;
+
+#. **inexistent**, if `c` never changed state.
+
+A ledger is consistent for `c` exactly if **Exercise** and **Fetch** actions on `c` happen only when `c` is active,
+and **Create** actions only when `c` is inexistent.
+The figures below visualize the state of different contracts at all points in the example ledger.
+
+.. https://www.lucidchart.com/documents/edit/19226d95-e8ba-423a-8546-e5bae6bd3ab7
+.. figure:: ./images/consistency-paint-offer-activeness.svg
+   :align: center
+   :width: 100%
+
+   Activeness of the `PaintOffer` contract
+
+.. https://www.lucidchart.com/documents/edit/19226d95-e8ba-423a-8546-e5bae6bd3ab7
+.. figure:: ./images/consistency-alice-iou-activeness.svg
+   :align: center
+   :width: 100%
+
+   Activeness of the `Iou Bank A` contract
+
+The notion of order can be defined on all the different ledger structures: actions, transactions, lists of transactions,
+and ledgers.
+Thus, the notions of (internal) activeness consistency, inputs and outputs, and contract state can also all be defined on all these
+structures.
+The **active contract set** of a ledger is the set of all contracts
+that are active on the ledger. For the example above, it consists
+of contracts `Iou Bank P` and `PaintAgree P A`.
+
+.. _da-model-key-consistency:
+
+Key consistency
+```````````````
+
+Contract keys introduce a key uniqueness constraint for the ledger.
+To capture this notion, the contract model must specify for every contract in the system whether the contract has a key and, if so, the key.
+Every contract can have at most one key.
+
+Like contracts, every key has a state.
+An action `act` is an **action on a key** `k` if 
+
+- `act` is a **Create**, **Exercise**, or a **Fetch** action on a contract `c` with key `k`, or
+- `act` is the key assertion **NoSuchKey** `k`.
+
+.. _def-key-state:
+  
+Definition »key state«
+  The **key state** of a key on a ledger is determined by the last action `act` on the key:
+
+  - If `act` is a **Create**, non-consuming **Exercise**, or **Fetch** action on a contract `c`,
+    then the key state is **assigned** to `c`.
+
+  - If `act` is a consuming **Exercise** action or a **NoSuchKey** assertion,
+    then the key state is **free**.
+
+  - If there is no such action `act`, then the key state is **unknown**.
+
+Key consistency ensures that there is at most one active contract for each key and that all key assertions are satisfies.
+
+.. _def-key-consistency:
+
+Definition »key consistency«
```

Comment:
Please make a note that this definition can be also be defined on actions, transactions and lists of transactions.

Reply:
Added a sentence at the end of the subsection.